### PR TITLE
Update API references for Callbacks

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Update API references for Callbacks

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @awaelchli @ananthsub @daniellepintz

### CODE REVIEW

File: docs/source/api_references.rst [hunk 1]
```diff
@@ -65,14 +65,27 @@ Callbacks API
...
+    :template: classtemplate.rst
 
-    base
-    early_stopping
-    gpu_stats_monitor
-    gradient_accumulation_scheduler
-    lr_monitor
-    model_checkpoint
-    progress
+    Callback
```
Comment (rohitgr7) [inline docs/source/api_references.rst:70]: LG! maybe order them alphabetically?

File: docs/source/api_references.rst [hunk 2]
```diff
@@ -65,14 +65,27 @@ Callbacks API
...
+    BackboneFinetuning
+    GPUStatsMonitor
+    GradientAccumulationScheduler
+    LambdaCallback
+    LearningRateMonitor
+    ModelCheckpoint
+    BasePredictionWriter
+    ModelPruning
+    QuantizationAwareTraining
+    RichModelSummary
```
Comment (rohitgr7) [inline docs/source/api_references.rst:83]: ModelSummary is missing. Maybe use this list just to ensure all are there: https://pytorch-lightning.readthedocs.io/en/latest/extensions/callbacks.html#built-in-callbacks

Comment (kaushikb11) [reply docs/source/api_references.rst:85]: Good catch, updated!

File: docs/source/api_references.rst (hunk 1 above)
Comment (kaushikb11) [reply docs/source/api_references.rst:73]: Updated! @rohitgr7

Comment (rohitgr7) [reply docs/source/api_references.rst:73]: or let's keep Callback at first since it's the base class, then all others a-z?

Comment (kaushikb11) [reply docs/source/api_references.rst:73]: That's inconsistent.

**Review by rohitgr7 — APPROVED** [review]

File: docs/source/api_references.rst (hunk 2 above)
Comment (ananthsub) [reply docs/source/api_references.rst:85]: Why do we have these in multiple spots? Doesn't that keep the risk of these falling out of sync?

File: docs/source/api_references.rst (hunk 1 above)
Comment (ananthsub) [reply docs/source/api_references.rst:73]: Should we move deprecated callbacks to the bottom? Or mark them as such on the docs page? Or will those automatically have the warning in the class docs?

Comment (rohitgr7) [reply docs/source/api_references.rst:73]: yes they will show up in docs:
Screenshot 2022-01-06 at 11 09 16 PM

**Review by awaelchli — APPROVED** [review]

File: docs/source/api_references.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source/api_references.rst:73]: if a concern is that alphabetic ordering is separating semantically related callbacks too much, you could also group them under categories. this is what we did for loops for example.

Comment (kaushikb11) [reply docs/source/api_references.rst:73]: I think the current list works. Deprecated Callbacks has the deprecation description.

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Comment by akihironitta** [comment]
Converting to draft as the build's failing...

**Review by tchaton — APPROVED** [review]
LGTM !

**Pull request merged**